### ISSUE TITLE
OrgTorch Fade activation - Feature request

### ISSUE THREAD

**Issue by Fa-b** [kolton/d2bot-with-kolbot#2173]
Stumbled over:
```
Config.OrgTorch.GetFade = false; // Get fade by standing in a fire. You MUST have Last Wish or Treachery on your character being worn.
```
Would it be to risky an attempt to check the inventory for `Treachery` armor and implement an automatic armor switch to get Fade as a prebuff?
Maybe have the armor located at a predefined position and have the inventory locked there for Pickit to ensure it does not interfere in any way.
Treachery is not really a sensible option for Uber Tristram.

**Comment by InterestingIndeed** [comment]
There is Autoequip in the charconfig, but I cant find anymore information about it. Maybe that could be used?

**Comment by Fa-b** [comment]
Where is that Autoequip property? Can't find it in my config

**Comment by 5noop** [comment]
Autoequip isn't in char config (anymore ?) but you can still use it, more info in pickit/Autoequip.

> Would it be to risky an attempt to check the inventory for `Treachery` armor and implement an automatic armor switch to get Fade as a prebuff?

It's always a bit risky because if you dc during the switch, the item on cursor will be dropped in next game. Then you have to manage these cases:
- what if you disconnect with treachery on ? need check for the right armor at game start ?
- treachery must not be muled but you shouldn't chose to not mule every treachery either
- you cannot select a fixed slot for the armor since you don't know how the player will manage his inventory/stash.

**Comment by Fa-b** [comment]
Found the AutoEquip property in Config.js:
```
// Experimental
FastParty: false,
AutoEquip: false,
```

And in Misc it is used:
```js
var Item = {
	hasTier: function (item) {
		return Config.AutoEquip && NTIP.GetTier(item) > 0;
	},
	canEquip: function (item) {
		[..]
	},
	// Equips an item and throws away the old equipped item
	equip: function (item, bodyLoc) {
		[..]
	},
	getEquippedItem: function (bodyLoc) {
		[..]
	},
	getBodyLoc: function (item) {
		[..]
	},
	autoEquipCheck: function (item) {
		[..]
	},
	// returns true if the item should be kept+logged, false if not
	autoEquip: function () {
		[..]
	}
};
```

I spared you details of function implementations. The comments make me guess this is implemented in favour of AutoBuild.js.
I guess there are some usable functions in there already for the purpose I have in mind!

@5noop The disconnect possibility is one of the risks I was thinking about. Of course the user should be made aware of that risk..
> what if you disconnect with treachery on ? need check for the right armor at game start ?

That shouldn't be too problematic to achieve. `getEquippedItem` func above might do exactly that. Would do it at the start of the script and only IF that option is configured though and not generally.

> treachery must not be muled but you shouldn't chose to not mule every treachery either

I would have no problem with removing `Treachery` from my mule list. My Smiter won't make runewords anyways.

> you cannot select a fixed slot for the armor since you don't know how the player will manage his inventory/stash.

I was thinking about a similar approach as for the general inventory configuration, for example:
```js
Config.OrgTorch.TreacheryLocation[0] = [1,1,0,0,0,0,0,0,0,0];
Config.OrgTorch.TreacheryLocation[1] = [1,1,0,0,0,0,0,0,0,0];
Config.OrgTorch.TreacheryLocation[2] = [1,1,0,0,0,0,0,0,0,0];
Config.OrgTorch.TreacheryLocation[3] = [0,0,0,0,0,0,0,0,0,0];
```

Of course, either you would have to Bitwise 'OR' these with the existing Inventory config, or copy these locations to there.

I agree the setup would be something for more advanced users (unless its builtin to the trunk branch online). But wouldn't the benefit be worth it? If you don't say it is an unnecessary addition I would go ahead and implement a solution (even test it on my own risk ;D).

**Comment by 5noop** [comment]
> unless its builtin to the trunk branch online

Yeah that's what my comments were about. I think it's a cool feature and if you want you should definitely try to implement your own solution.

I just meant there are problems to solve before attempting to write a clean version that any user could run with minimal configuration. Ideally it would be something like that:
`Config.OrgTorch.GetFade = false; // Get fade by standing in a fire. If you are not wearing Last Wish or Treachery, it will look for a Treachery in your stash and use it to prebuff.`

**Comment by unseenstorm** [comment]
You think we could add a pickit line in town just in case it gets stuck on the cursor and dropped in the new game?

**Comment by Fa-b** [comment]
> You think we could add a pickit line in town just in case it gets stuck on the cursor and dropped in the new game?
Yes, and thanks for the suggestion. There are several optimizations needed for this to be 'production ready'. Dropping the item when inv, stash or cube is full is also not a really clever solution :). Right now I only implemented Treachery also since I though Last Wish would not need switch, but who am I to make that decision for you. Also it will still go getFade if you already Last Wish in one of your hands. I just made myself a Last Wish only for the purpose of testing as many possible situations that I can think of right now. Loosing a Treachery would not drive me too mad, my switch is a 15ed ap enigma though :x.
Right now, I have had my smiter get 8x organ sets and 8x torches respectively using the armor switch above. No failure so far... haven't tested anything 'exotic' though. I think the least risky method is to have the switch armor in stash. Can't think of a situation where the original space in stash will be filled between the switches.
Please more suggestions like that.

**Comment by bobbyjose1** [comment]
I'm very excited to test out this script! When I copy and replace this.getFade with your script, however, it gives me this error. Please help! I tried downloading a fresh version of kolbot and I'm still having the same problem. I checked it a million times to make sure i copied and pasted it right. I literally delete the old this.getFade function entirely and replace it with the code you provided. The script works just fine before editing it and after I replace this.getFade, it gives me this error.
![Error message](https://user-images.githubusercontent.com/3161788/85247564-0af5bd80-b403-11ea-8e71-f80412c60dd7.png)

**Comment by Fa-b** [comment]
probably smt went wrong with the encoding during copy paste.. I can only assume it might have to do with one special character used in print statements `�` which should be `ÿ`.
What editor did you use to paste the code into? does your browser display ÿ proplerly? Maybe try a different browser.

**Comment by bobbyjose1** [comment]
![Error message 2](https://user-images.githubusercontent.com/3161788/85484234-44e9cf80-b57b-11ea-91b5-bd9bd79bbae4.png)

Thanks for the reply! This is how it looks for me in notepad++. I commented out all the lines that have ÿ in it and my bot is able to load OrgTorch now.  Obviously this doesn't fix the problem but it looks like we narrowed it down.

**Comment by Fa-b** [comment]
good, not even sure that string colorcode replacement is necessary tbh.. it's anyways just debugging messages so should be fine to comment out.

**Comment by bobbyjose1** [comment]
Cool script Fa-b thanks! You are right. It works just fine now.

**Comment by Fa-b** [comment]
Please don‘t underestimate the risk of loosing an armor though! This is still first draft and for sure needs improvement!

**Comment by bobbyjose1** [comment]
Works great! I have farmed a lot of torches no problem.. only thing i had to change was the location he stands in the fire.. sometimes it would work and sometimes it wouldn't because the pathing isn't the same as teleporting. I changed the coordinates from "Pather.moveTo(7811, 5872);  // move into fire" to "Pather.moveTo(7812, 5870);  // move into fire" and it works fine every time. I was thinking about doing the things already mentioned to make the bot safer but don't think its worth the time with the amount of torches i've farmed with no loss of items.

**Comment by JamesHatter** [comment]
Possible to just swap armors with merc so no inventory space is required?  I’d try it as is to answer my own question but I don’t have a treach yet. Just started back up with the ladder reset.

**Comment by exiled-one** [comment]
Where are these item prefixes being used in the item.getIndex() function coming from? I'm trying to find the one for Heart Of The Oak but I'm not seeing the file they are in
